sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IO.java

```diff
@@ -472,24 +548,120 @@ public void initClient() throws IOException {
       this.client = new HttpHealthcareApiClient();
     }
 
+    @GetInitialRestriction
+    public OrderedTimeRange getEarliestToLatestRestriction(@Element String hl7v2Store)
+        throws IOException {
+      from = this.client.getEarliestHL7v2SendTime(hl7v2Store, this.filter);
+      // filters are [from, to) to match logic of OffsetRangeTracker but need latest element to be
+      // included in results set to add an extra ms to the upper bound.
+      to = this.client.getLatestHL7v2SendTime(hl7v2Store, this.filter).plus(1);
+      return new OrderedTimeRange(from, to);
+    }
+
+    @NewTracker
+    public OrderedTimeRangeTracker newTracker(@Restriction OrderedTimeRange timeRange) {
+      return timeRange.newTracker();
+    }
+
+    @SplitRestriction
+    public void split(
+        @Restriction OrderedTimeRange timeRange, OutputReceiver<OrderedTimeRange> out) {
+      // TODO(jaketf) How to pick optimal values for desiredNumOffsetsPerSplit ?
```

Comment:
Typically, doing one split for every 64mbs of output has been our guidance here in the past.

Dynamic splitting is meant to fill in the gap if there is too little splitting or a specific restriction has a lot more data then other restrictions.

Reply:
Unfortunately, in this use case dynamic splitting would be crucial because we can't know the distribution of data in the restriction dimension (sendTime). 

If you imagine a hospital might be much busier during daytime / weekdays than night times weekends (though never dormant due to ICU and emergency services). "Day time" might change base on hospital location, week days are subject to holidays, etc.

Data distribution in sendTime may be subject to significant spikes if one of the upstream systems populating sendTime has to backfill after a maintenance period and doesn't responsibly set this field to event time but sets all of the sendTimes to a short range of  backfill 
processing time (this is sub optimal behavior of that system but sometimes a reality).

Reply:
Yeah, you'll have to choose a reasonable value based upon your knowledge within this domain and I understand that this may not work in practice.

In the case where people send a lot of data for a narrow time range, you would have to support filtering based upon other properties such as messageType or sendFacility and could break up the space based upon prefixes such as `~a`, `~b`, `NOT ~a AND NOT ~b` but this is dependent on this being efficient in the service as well.

Reply:
Yeah I think the "spiky backfill" (many cases in a small sendTime) is a corner case of a hot split that would just be slow and users would have to accept that or take it up with their upstream system.

splitting on messageType / sendFacility are probably more popular logical filters and feels like a hack for a corner case that might mess with performance under the "typical" distribution of data in sendTime.

Reply:
My intuition says hourly splits seems like a more reasonable starting point than day but I'm unsure about this without testing on a few datasets.

However for super large time ranges 10 years = 87,600 splits. naturally this explodes if we think about even smaller time ranges.

For example, I assume the initial collection of splits returned in SplitRestriction has to fit in memory? 
Or even just the loop to define all these might take a significant amount of time

Reply:
That seems like a lot.

Dataflow has an API limit of 20mbs for split descriptions when being returned which usually tops out around 10k splits for sources but even 10k is too much. Typically 20-50 splits is enough since dynamic splitting will ramp that up to 1000s if necessary.

Reply:
Interesting, thanks for those details! 

But I thought Dataflow doesn't really do dynamic splitting yet so the initial split is all that matters? Maybe I'm confusing dynamic rebalancing != dynamic splitting?
  
Perhaps daily is the appropriate happy medium for now? 
1000s not 10,000s of initial splits.

In the future when dynamic splitting is supported bump initial split default to weeks?

Reply:
dynamic splitting is a subset of dynamic rebalancing since dynamic rebalancing also covers other aspects such as splitting a GroupByKey, scaling up/down workers, ...

dynamic splitting is supported by batch Dataflow for bounded sinks using runner v1 and splittable dofns when running a batch pipeline when using runner v2

Reply:
Ok I think that dynamic splitting is "nice to have" for this connector and can wait for v2 connector.

Just adding initial splitting w/ SDF and v1 runner is a huge improvement over the existing implementation.